### PR TITLE
Backport recovery TAP tests

### PR DESCRIPTION
**Pull request by roicos**
This PR backports part of recovery TAP tests that can be supported on GPDB6 and related TAP framework changes.
See the individual commits for more information.

Note: this PR enables utility mode hot_standby strictly for testing only.

### CODE REVIEW

**Review by jimmyyih — APPROVED** [review]
Looks fine to me.  The only concern I have is enabling utility mode hot standby in 6X_STABLE but that should be okay as long as we strictly convey that it's currently only for developer testing usage only and not to be used in production systems.

**Review by huansong — APPROVED** [review]
LGTM!

File: src/test/perl/PostgresNode.pm [hunk 1]
```diff
@@ -818,6 +818,10 @@ sub promote
 	print "### Promoting node \"$name\"\n";
 	TestLib::system_or_bail('pg_ctl', '-D', $pgdata, '-l', $logfile,
 							'promote');
+	# GPDB: In 6X_STABLE, pg_ctl promote does not wait until promotion is finished
+	# so we need to manually check in a loop.
+	$self->poll_query_until('postgres', "SELECT NOT pg_is_in_recovery()")
+		or die "Timed out while waiting for standby to promote";
```
Comment (soumyadeep2007) [inline src/test/perl/PostgresNode.pm:825]: You mean if you pass `-w` to the `pg_ctl` command above, it will not honor it?

**Review by soumyadeep2007 — APPROVED** [review]
Please mention commit SHA which you backported from for the following commits:

[Allow read-only connections on a hot standby](https://github.com/greenplum-db/gpdb/pull/13375/commits/1edab2b26a3aad0b6ffea4db0489cc5697d8b7ae)
[Allow pg_ctl kill to send SIGKILL.](https://github.com/greenplum-db/gpdb/pull/13375/commits/c8c982e111871f5eb6ebacc1d08d3dc3d86411d7)

Awesome work! Great detail in the commit messages, `GPDB :` comments which are always nice to see and so crucial for such backports. :clap:

File: src/backend/postmaster/postmaster.c [hunk 2]
```diff
@@ -2618,6 +2618,18 @@ ProcessStartupPacket(Port *port, bool SSLdone)
 				break;
 			}
 
+			/*
+			 * Allow connections if hot_standby is on and our postmaster is
+			 * acting as a standby.
+			 *
+			 * GPDB_12_MERGE_FIXME: checking a GUC is not a good idea here.
```
[the remark this answers is not in the document]
Comment (ashwinstar) [reply src/backend/postmaster/postmaster.c:2625]: need to address this comment

[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/backend/postmaster/postmaster.c:2625]: Can put something like this:
```
GPDB_XX_MERGE_FIXME: This is backported from master (GPDB 7) strictly for developer testing usage only and should not be used in production systems.  Checking the GUC here is generally not a good idea.
```

File: src/test/perl/PostgresNode.pm [hunk 3]
```diff
@@ -416,7 +433,7 @@ sub init
 		print $conf "shared_buffers = 1MB\n";
 		print $conf "wal_log_hints = on\n";
 		print $conf "hot_standby = on\n";
-		print $conf "max_connections = 10\n";
+		print $conf "max_connections = 20\n";
```
[the remark this answers is not in the document]
Comment (roicos) [reply src/test/perl/PostgresNode.pm:437]: This is a part of backporting of https://github.com/greenplum-db/gpdb-postgres-merge/commit/18ed14267da204506a253e12963957686a991479 and is there in master. From observations, it's needed to make some of the tests work.

File: src/test/perl/PostgresNode.pm [hunk 4]
```diff
@@ -140,6 +140,8 @@ sub new
 	my $self = {
 		_port    => $pgport,
 		_host    => $pghost,
+		# GPDB needs dbid for a node for certain operations
+		_dbid    => int(rand(10)),
```
[the remark this answers is not in the document]
Comment (roicos) [reply src/test/perl/PostgresNode.pm:144]: This is a part of backporting of https://github.com/greenplum-db/gpdb-postgres-merge/commit/18ed14267da204506a253e12963957686a991479. It's been refactored in master but we are just going for the minimum backports without any additional changes.

File: src/test/perl/PostgresNode.pm (hunk 1 above)
Comment (roicos) [reply src/test/perl/PostgresNode.pm:825]: There is no -w option in Postgres 9.4 pg_ctl promote. It was added later.

**Pull request merged**